File: courses/codes/mst_prim_solution.py
# Prim's algorithm
# Input: graph G (adj list), weights dict (keys are edges, which are tuples; values are weights)
# Output: Set of edges that make up the MST
def prim(G, W):
    T_nodes, T_edges = set(), set()                             # the min spanning tree
    V = dict(zip(G.keys(), [False]*len(G.keys())))              # booleans for visited nodes

    s = list(G.keys())[0]                                       # source node (arbitrary chosen)
    T_nodes.add(s)                                              # add node to the MST
    V[s] = True                                                 # node has been visited

    while len(T_edges) < len(G) - 1:

        # obtain the connected nodes and connected edges (i.e., the candidate edges
        # that will be added to the MST)
        connected_nodes, connected_edges = set(), set()
        for node in T_nodes:
            nbrs = G[node]
            for nbr in nbrs:
                if not V[nbr]:
                    connected_nodes.add(nbr)

        for n1 in T_nodes:
            for n2 in connected_nodes:
                key = (n1, n2) if n1 < n2 else (n2, n1)
                if key in W:
                    connected_edges.add(key)

        # among the candidate edges, get the cheapest weight
        min_weight = float('inf')
        min_edge = None
        for edge in connected_edges:
            weight_key = (edge[0], edge[1]) if edge[0] < edge[1] else (edge[1], edge[0])
            if W[weight_key] < min_weight:
                min_weight = W[weight_key]
                min_edge = edge

        # add the cheapest weight to the MST
        # then identify the new nodes in the MST as visited
        e1, e2 = min_edge
        V[e1], V[e2] = True, True
        T_edges.add(min_edge)
        T_nodes = T_nodes.union({e1, e2})

    return T_nodes, T_edges

# Helper Functions Only =============================================

# ensures that (e1, e2) edges are e1 < e2
def ordered_pairs(W):
    new_W = dict()
    for k, v in W.items():
        if k[0] > k[1]:
            new_key = (k[1], k[0])
            new_W[new_key] = v
        else:
            new_W[k] = v
    return new_W

# return the graph adj list G given the weights dict
def get_graph_given_weights(W):
    G = dict()
    for w in W:
        if w[0] not in G:
            G[w[0]] = []
        G[w[0]].append(w[1])

        if w[1] not in G:
            G[w[1]] = []
        G[w[1]].append(w[0])
    return G




n = int(input())
W = dict()
for _ in range(n):
    arr = input().split(' ')
    W[(arr[0], arr[1])] = int(arr[2])

W = ordered_pairs(W)            # ensures that each edge (e1, e2) is e1 < e2 to avoid duplicates
G = get_graph_given_weights(W)
MST_Nodes, MST_Edges = prim(G, W)

for e in MST_Edges:
    print(e[0] + ' ' + e[1])

File: courses/codes/cclv1-sr-class/MSTPrimSolution.java
import java.io.*;
import java.util.*;

public class MSTPrimSolution {

	public static void main(String[] args) {
		
		ex2();
    }

	
	
	private static void ex2() {
		Scanner sc = new Scanner(System.in);
		Map<String[], Integer> W = new HashMap<>();
		int n = Integer.parseInt(sc.nextLine());
		for (int i = 0; i < n; i++) {
			String[] arr = sc.nextLine().split(" ");
			Integer w = Integer.parseInt(arr[2]);
			String[] edge = {arr[0], arr[1]};
			W.put(edge, w);
		}
		W = orderedPairs(W);
		Map<String, List<String>> G = getGraphFromWeights(W);
		Set<String[]> MSTEdges = prim(G, W);

        System.out.println("Minimum Spanning Tree Edges:");
        for (String[] edge : MSTEdges) {
            System.out.println(Arrays.toString(edge));
        }

        sc.close();
	}

	private static Set<String[]> prim(Map<String, List<String>> G, Map<String[], Integer> W) {
	    Set<String> TNodes = new HashSet<>();
	    Set<String[]> TEdges = new HashSet<>();
	    Map<String, Boolean> visited = new HashMap<>();
	    for (String node : G.keySet()) {
	        visited.put(node, false);
	    }

	    List<String> nodes = new ArrayList<>(G.keySet());
	    String s = nodes.get(0);
	    TNodes.add(s);
	    visited.put(s, true);

	    while (TEdges.size() < G.size() - 1) {
	        Set<String> connectedNodes = new HashSet<>();
	        Set<String[]> connectedEdges = new HashSet<>();
	        for (String node : TNodes) {
	            List<String> nbrs = G.get(node);
	            for (String nbr : nbrs) {
	                if (!visited.get(nbr)) {
	                    connectedNodes.add(nbr);
	                }
	            }
	        }
	        
	        for (String n1 : TNodes) {
	            for (String n2 : connectedNodes) {
	                String[] key = new String[2];
	                if (n1.compareTo(n2) > 0) {
	                		key[0] = n2;
	                		key[1] = n1;
		            } else {
		            		key[0] = n1;
		            		key[1] = n2;
		            }
	                
	                for (String[] k : W.keySet()) {
	                		if (k[0].equals(key[0]) && k[1].equals(key[1])) {
	                			connectedEdges.add(key);
	                			break;
	                		}
	                }
	            }
	        }

	        int minWeight = Integer.MAX_VALUE;
	        String[] minEdge = {null, null};
	        for (String[] edge : connectedEdges) {
	        		String[] kEdge = new String[2];
	        		Integer vWeight = 0;
	        		for (Map.Entry<String[],Integer> entry : W.entrySet())   {
	            		String[] myEdge = entry.getKey();
	            		boolean b1 = edge[0].equals(myEdge[0]);
	            		boolean b2 = edge[1].equals(myEdge[1]);
	            		if (b1 && b2) {
	            			kEdge[0] = myEdge[0];
	            			kEdge[1] = myEdge[1];
	            			vWeight = entry.getValue();
	            			break;
	            		}
	            }
	        		if (vWeight < minWeight) {
	        			minWeight = vWeight;
	        			minEdge[0] = kEdge[0];
	        			minEdge[1] = kEdge[1];
	        		}
	        }
	     

	        String e1 = minEdge[0];
            String e2 = minEdge[1];
            visited.put(e1, true);
            visited.put(e2, true);
            
            TEdges.add(minEdge);
            TNodes.add(e1);
            TNodes.add(e2);
	    }

	    return TEdges;
	}


	// Anything below this are helper functions only
	// DO NOT modify!

    private static Map<String[], Integer> orderedPairs(Map<String[], Integer> W) {
        Map<String[], Integer> orderedWeights = new HashMap<>();
        for (String[] k : W.keySet()) {
        		String[] newKey = new String[2];
            if (k[0].compareTo(k[1]) > 0) {
            		newKey[0] = k[1];
            		newKey[1] = k[0];
            		orderedWeights.put(newKey, W.get(k));
            }
            orderedWeights.put(k, W.get(k));
        }
        return orderedWeights;
    }


    
    private static Map<String, List<String>> getGraphFromWeights(Map<String[], Integer> W) {
   		Map<String, List<String>> G = new HashMap<>();
        for (Map.Entry<String[], Integer> entry : W.entrySet()) {
        		String e1 = entry.getKey()[0];
        		String e2 = entry.getKey()[1];
        		Integer weight = entry.getValue();
        		if (!G.containsKey(e1)) {
        			List<String> L1 = new ArrayList<>();
        			G.put(e1, L1);
        		}
        		List<String> l1 = G.get(e1);
        		l1.add(e2);
        		Set<String> s1 = new HashSet<>(l1);
        		l1 = new ArrayList<>(s1);
        		G.put(e1, l1);
        		if (!G.containsKey(e2)) {
        			List<String> L2 = new ArrayList<>();
        			G.put(e2, L2);
        		}
        		List<String> l2 = G.get(e2);
        		l2.add(e1);
        		Set<String> s2 = new HashSet<>(l2);
        		l2 = new ArrayList<>(s2);
        		G.put(e2, l2);
        }
        return G;
    }
    
   
}
